### ISSUE TITLE
Support async thread entry points

### ISSUE THREAD

**Issue by chemicstry** [chemicstry/wasm_thread#10]
Async entry points would allow using async channels inside web worker threads.

This feature would probably need a major rewrite as discussed in https://github.com/chemicstry/wasm_thread/issues/8

**Comment by Davidster** [comment]
Hello! I am currently encountering the same issue as #6 and am wondering if there might be a way to fix the issue without the need for supporting postMessage? Currently I'm planning on doing something like this:

```rust
fn spawn_thread<F, T>(f: F)
where
    F: FnOnce() -> T,
    F: Send + 'static,
    T: Send + 'static,
{
    if cfg!(target_arch = "wasm32") {
        thread::spawn(|| {
            f();
            wasm_bindgen::throw_str("Cursed hack to keep workers alive. See https://github.com/rustwasm/wasm-bindgen/issues/2945");
        });
    } else {
        thread::spawn(f);
    }
}
```

At a minimum it is possible to make it work with the error. I don't see what exactly is stopping wasm_thead from being able to block in worker threads, although I admit I haven't read through the crate's code.

**Comment by chemicstry** [comment]
@Davidster I would rather avoid hacks with throws.

So I looked more into this issue and found something stupid simple, the web worker script has a close call after executing closure: https://github.com/chemicstry/wasm_thread/blob/main/src/web_worker.js#L25C11-L25C11, could you test if removing it makes your usecase work? You might need to also remove it from `web_worker_module.js` if you are using ES modules.

I tested it locally and it seems to work, maybe everyone missed this `close()` call when discussing the previous issue

**Comment by Davidster** [comment]
Yes, after posting my comment I read more code and found the close() call. Removing it might indeed fix the issue and I will test it in a few days when I get the chance. I was worried that the close() call was needed to prevent a leak or something, but the browser should be able to clean up workers automatically when they have no more pending tasks, right? I wonder if there's a way I can test it, maybe by spawning a lot of wasm_thread's and checking in the dev tools window to see how many active workers stay alive (not sure if such a tool exists).

Thank you for your reply!

**Comment by chemicstry** [comment]
In theory, web workers should terminate automatically when there are no items left in worker's event queue (pending async promises) and main thread has dropped worker handle (can no longer `postMessage()`). However, I could not replicate this behavior yet. Maybe there is still some dangling reference left that prevent garbage collector cleanup.

**Comment by adamgerhant** [comment]
I just ran into a very similar issue when running async code with futures_lite::future::block_on in a web worker. The web worker would close before a promise could resolve. I think it might be worth removing the .close(), or at least making it an option and documenting the fact that it not work with any async code in the web worker.

**Comment by chemicstry** [comment]
@adamgerhant could you test it on a new refactored v0.3.0 release? We now have a test for this behavior: https://github.com/chemicstry/wasm_thread/blob/49ae81d67601bdc2e64d8ef96e14c7e62107aa4c/tests/wasm.rs#L102

**Comment by adamgerhant** [comment]
I tried it with v0.3.0 and still no luck. Removing the close() function also didn't fix it. For reference here is the broken wasm_thread usage versus my custom implementation 

 ## Wasm_thread not working: "created thread adapter" never gets printed
```
 wasm_thread::spawn(|| {
         futures::executor::block_on(async move {
            let instance = wgpu::Instance::default();
            log::info!("created thread instance");
            let thread_adapter = instance
                .request_adapter(&wgpu::RequestAdapterOptions::default())
                .await
                .unwrap();
            log::info!("created thread adapter"); //not printed
         })
     });
```

## Working thread entry point
```
#[wasm_bindgen]
pub async fn thread_entry_point() {
    let instance = wgpu::Instance::default();
    log::info!("created thread instance");
    let thread_adapter = instance
        .request_adapter(&wgpu::RequestAdapterOptions::default())
        .await
        .unwrap();
    log::info!("created thread adapter"); //printed
}

I believe the reason wasm_thread does not work is because .request_adapter [internally awaits](https://github.com/gfx-rs/wgpu/blob/152a94bc6c502226d9871f28e35db0b755ea35bf/wgpu/src/lib.rs#L1912) a [jsFuture](https://github.com/gfx-rs/wgpu/blob/152a94bc6c502226d9871f28e35db0b755ea35bf/wgpu/src/backend/webgpu.rs#L1143) based on a js_sys promise, and block_on does not work with jsFutures. 

Here is a minimal example to demonstrate:
```
pub fn create_promise() -> Promise {
    let promise = Promise::new(&mut |resolve, reject| {
        log::info!("promise");
        let success = true;
        if success {
            resolve.call0(&JsValue::null()).unwrap();
        } else {
            let error_msg = JsValue::from_str("An error occurred");
            reject.call1(&JsValue::null(), &error_msg).unwrap();
        }
    });

    promise
}

#[wasm_bindgen]
pub async fn main(){
    console_log::init().expect("could not initialize logger");
    log::info!("before");
    let js_future = wasm_bindgen_futures::JsFuture::from(create_promise());
    js_future.await;
    log::info!("after");
    
    wasm_thread::spawn(|| {
        futures::executor::block_on(async move {
            log::info!("before");
            let js_future = wasm_bindgen_futures::JsFuture::from(create_promise());
            js_future.await;
            log::info!("after");  //after is not printed
        })
    });
}
```
Output:
before
promise         (awaiting jsFuture from main thread works)
after
before 
promise          (awaiting jsFuture promise with block_on does not resume execution after the promise)

Ultimately it would be great to rework wasm_thread to not require block_on for async functions, but I'm not sure if that would be possible.

**Comment by chemicstry** [comment]
Yeah the problem here is that `block_on` parks the thread (possibly on atomic wait) and javascript event loop cannot run while the thread is parked. The close call is needed in order to not leak memory when threads are actively being created/detroyed. I will think how this could be fixed without deviating too much from the std thread API.

**Comment by adamgerhant** [comment]
I added a spawn_async method that spawns the future with spawn_local, and uses a static arc flag that is updated when the function is complete and then used to close the thread. It works well so far, and I'll try create a pull request today. The main problem so far is that I wasnt able to get the join handle properly implemented, but I think that should be possible.
